File: web/src/api/project-repositories.ts
import { etagForVersion, requireVersioned, v1Delete, v1Get, v1Post } from './v1/client'

export interface ProjectRepository {
  id: string
  canonical_web_url: string
	label: string
	provider: 'gitlab' | 'github'
	origin: string
	provider_repository_id: string
  path_with_namespace: string
  default_branch: string
  bound_at: string
}

export interface ProjectRepositoryMutation {
  project_version: number
  repository: ProjectRepository
}

export function listProjectRepositories(projectNumber: number): Promise<ProjectRepository[]> {
  return v1Get<{ items: ProjectRepository[] }>(
    `/api/v1/projects/${projectNumber}/repositories`,
  ).then(({ value }) => value.items)
}

export function bindProjectRepository(
  projectNumber: number,
  projectVersion: number,
  repositoryURL: string,
): Promise<ProjectRepositoryMutation> {
  return v1Post<ProjectRepositoryMutation>(`/api/v1/projects/${projectNumber}/repositories`, {
    ifMatch: etagForVersion(projectVersion),
    body: { repository_url: repositoryURL },
  }).then((response) => requireVersioned(response).value)
}

export function unbindProjectRepository(
  projectNumber: number,
  projectVersion: number,
  repositoryID: string,
): Promise<ProjectRepositoryMutation> {
  return v1Delete<ProjectRepositoryMutation>(
    `/api/v1/projects/${projectNumber}/repositories/${repositoryID}`,
    { ifMatch: etagForVersion(projectVersion) },
  ).then((response) => requireVersioned(response).value)
}

File: internal/api/v1/project_repositories.go
package v1

import (
	"context"
	"fmt"
	"net/url"

	baseapi "github.com/wolfhead/pactline/internal/api"
	generated "github.com/wolfhead/pactline/internal/api/v1generated"
	"github.com/wolfhead/pactline/internal/store"
)

func (h *Handler) ListProjectRepositories(
	ctx context.Context,
	params generated.ListProjectRepositoriesParams,
) (generated.ListProjectRepositoriesRes, error) {
	subject, err := accessSubject(ctx)
	if err != nil {
		return nil, err
	}
	repositories, err := h.ProjectRepositories.List(ctx, params.Number, subject)
	if err != nil {
		return nil, err
	}
	items := make([]generated.ProjectRepository, len(repositories))
	for index, repository := range repositories {
		items[index], err = projectRepositoryFromDomain(repository)
		if err != nil {
			return nil, err
		}
	}
	return &generated.ProjectRepositoryListHeaders{
		XRequestID: generated.NewOptString(baseapi.RequestIDFromContext(ctx)),
		Response:   generated.ProjectRepositoryList{Items: items},
	}, nil
}

func (h *Handler) BindProjectRepository(
	ctx context.Context,
	req *generated.ProjectRepositoryBind,
	params generated.BindProjectRepositoryParams,
) (generated.BindProjectRepositoryRes, error) {
	expectedVersion, err := parseIfMatch(params.IfMatch)
	if err != nil {
		return nil, err
	}
	operation, _, err := operationContext(ctx)
	if err != nil {
		return nil, err
	}
	subject, err := accessSubject(ctx)
	if err != nil {
		return nil, err
	}
	mutation, err := h.ProjectRepositories.Bind(
		ctx, params.Number, expectedVersion, req.RepositoryURL.String(), subject, operation,
	)
	if err != nil {
		return nil, err
	}
	return projectRepositoryMutationResponse(ctx, mutation)
}

func (h *Handler) UnbindProjectRepository(
	ctx context.Context,
	params generated.UnbindProjectRepositoryParams,
) (generated.UnbindProjectRepositoryRes, error) {
	expectedVersion, err := parseIfMatch(params.IfMatch)
	if err != nil {
		return nil, err
	}
	operation, _, err := operationContext(ctx)
	if err != nil {
		return nil, err
	}
	subject, err := accessSubject(ctx)
	if err != nil {
		return nil, err
	}
	mutation, err := h.ProjectRepositories.Unbind(
		ctx, params.Number, expectedVersion, params.RepositoryID, subject, operation,
	)
	if err != nil {
		return nil, err
	}
	return projectRepositoryMutationResponse(ctx, mutation)
}

func projectRepositoryMutationResponse(
	ctx context.Context,
	mutation store.ProjectRepositoryMutation,
) (*generated.ProjectRepositoryChangedHeaders, error) {
	repository, err := projectRepositoryFromDomain(mutation.Repository)
	if err != nil {
		return nil, err
	}
	return &generated.ProjectRepositoryChangedHeaders{
		Etag:       generated.NewOptString(formatETag(mutation.ProjectVersion)),
		XRequestID: generated.NewOptString(baseapi.RequestIDFromContext(ctx)),
		Response: generated.ProjectRepositoryMutation{
			ProjectVersion: mutation.ProjectVersion,
			Repository:     repository,
		},
	}, nil
}

func projectRepositoryFromDomain(
	item store.ProjectRepositoryWithConnection,
) (generated.ProjectRepository, error) {
	canonicalURL, err := url.Parse(item.Connection.CanonicalWebURL)
	if err != nil {
		return generated.ProjectRepository{}, fmt.Errorf("parse canonical GitLab repository URL: %w", err)
	}
	origin, err := url.Parse(item.Connection.Origin)
	if err != nil {
		return generated.ProjectRepository{}, fmt.Errorf("parse GitLab origin: %w", err)
	}
	return generated.ProjectRepository{
		ID: item.Repository.ID, CanonicalWebURL: *canonicalURL,
		Label: item.Connection.Label, Origin: *origin,
		GitlabProjectID:   item.Connection.GitLabProjectID,
		PathWithNamespace: item.Connection.PathWithNamespace,
		DefaultBranch:     item.Connection.DefaultBranch, BoundAt: item.Repository.BoundAt,
	}, nil
}
